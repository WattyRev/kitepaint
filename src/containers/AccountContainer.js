import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { makeCancelable } from "../utils";
import { getUser } from "../redux/modules/user";
import {
  CHANGE_EMAIL,
  CHANGE_PASSWORD,
  DELETE_ACCOUNT
} from "../redux/actions";

export class AccountContainer extends React.Component {
  static propTypes = {
    /**
     * A function that renders content
     */
    children: PropTypes.func.isRequired,
    /**
     * Details about the current user. Provided by redux.
     */
    user: PropTypes.shape({
      email: PropTypes.string,
      id: PropTypes.string,
      isLoggedIn: PropTypes.bool,
      isLoggingIn: PropTypes.bool,
      username: PropTypes.string
    }),
    /**
     * Called when a change of password is requested. Provided by redux.
     */
    onEmailChange: PropTypes.func.isRequired,
    /**
     * Called when a change of email is requested. Provided by redux.
     */
    onPasswordChange: PropTypes.func.isRequired,
    /**
     * Called when account deletion is requested. Provided by redux.
     */
    onDeleteAccount: PropTypes.func.isRequired
  };

  state = {
    confirmNewPassword: "",
    currentPassword: "",
    deleteError: null,
    editingEmail: false,
    editingPassword: false,
    email: "",
    emailError: null,
    newPassword: "",
    passwordError: null
  };

  handleToggleEditEmail = () => {
    this.setState({
      emailError: null,
      editingEmail: !this.state.editingEmail,
      email: this.props.user.email
    });
  };

  handleToggleEditPassword = () => {
    this.setState({
      editingPassword: !this.state.editingPassword,
      confirmNewPassword: "",
      currentPassword: "",
      newPassword: "",
      passwordError: null
    });
  };

  handleEmailChange = email => {
    this.setState({
      emailError: null,
      email
    });
  };

  handleCurrentPasswordChange = currentPassword => {
    this.setState({
      currentPassword,
      passwordError: null
    });
  };

  handleNewPasswordChange = newPassword => {
    this.setState({
      newPassword,
      passwordError: null
    });
  };

  handleConfirmNewPasswordChange = confirmNewPassword => {
    this.setState({
      confirmNewPassword,
      passwordError: null
    });
  };

  handleSubmitEmail = event => {
    event.preventDefault();
    const request = makeCancelable(
      this.props.onEmailChange(this.props.user.id, this.state.email)
    );
    this._requests.push(request);
    request.promise
      .then(() => {
        this.setState({
          editingEmail: false
        });
      })
      .catch(message => {
        this.setState({
          emailError: message
        });
      });
  };

  handleSubmitPassword = event => {
    event.preventDefault();
    const request = makeCancelable(
      this.props.onPasswordChange({
        username: this.props.user.username,
        currentPassword: this.state.currentPassword,
        newPassword: this.state.newPassword,
        confirmNewPassword: this.state.confirmNewPassword
      })
    );
    this._requests.push(request);
    request.promise
      .then(() => {
        this.setState({
          editingPassword: false
        });
      })
      .catch(message => {
        this.setState({
          passwordError: message
        });
      });
  };

  handleDeleteAccount = password => {
    const request = makeCancelable(
      this.props.onDeleteAccount(this.props.user.id, password)
    );
    this._requests.push(request);
    request.promise.catch(message => {
      this.setState({
        deleteError: message
      });
    });
  };

  _requests = [];

  componentWillUnmount() {
    this._requests.forEach(request => request.cancel());
  }

  render() {
    return this.props.children({
      actions: {
        changeEmail: this.handleEmailChange,
        changeCurrentPassword: this.handleCurrentPasswordChange,
        changeNewPassword: this.handleNewPasswordChange,
        changeConfirmNewPassword: this.handleConfirmNewPasswordChange,
        toggleEditEmail: this.handleToggleEditEmail,
        toggleEditPassword: this.handleToggleEditPassword,
        submitEmail: this.handleSubmitEmail,
        submitPassword: this.handleSubmitPassword,
        deleteAccount: this.handleDeleteAccount
      },
      props: {
        confirmNewPassword: this.state.confirmNewPassword,
        currentPassword: this.state.currentPassword,
        deleteError: this.state.deleteError,
        editingEmail: this.state.editingEmail,
        editingPassword: this.state.editingPassword,
        email: this.state.editingEmail
          ? this.state.email
          : this.props.user.email,
        emailError: this.state.emailError,
        newPassword: this.state.newPassword,
        passwordError: this.state.passwordError,
        user: this.props.user
      }
    });
  }
}

const mapStateToProps = state => ({
  user: getUser(state)
});

const mapDispatchToProps = {
  onEmailChange: CHANGE_EMAIL,
  onPasswordChange: CHANGE_PASSWORD,
  onDeleteAccount: DELETE_ACCOUNT
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AccountContainer);
