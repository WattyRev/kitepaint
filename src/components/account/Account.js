import React from "react";
import styled from "styled-components";
import { Redirect } from "react-router-dom";
import AccountContainer from "../../containers/AccountContainer";
import {
  Button,
  Error,
  H1,
  H2,
  ModalPrompt,
  P,
  Spacer,
  Tile
} from "../../theme";
import ChangeEmail from "./ChangeEmail";
import ChangePassword from "./ChangePassword";

const StyleWrapper = styled.div`
  max-width: 1280px;
  padding: 16px;
  margin: 0 auto;

  .form-group {
    padding: 8px;
  }
`;

/**
 * A coordinating component that builds up the Account page.
 */
const Account = () => (
  <StyleWrapper>
    <AccountContainer>
      {accountData => (
        <React.Fragment>
          {!accountData.props.user.id && <Redirect to="/" />}
          <H1>Account Settings - {accountData.props.user.username}</H1>
          <Spacer top="md" bottom="md">
            <Tile>
              <ChangeEmail
                editing={accountData.props.editingEmail}
                email={accountData.props.email}
                error={accountData.props.emailError}
                onChangeEmail={accountData.actions.changeEmail}
                onSubmit={accountData.actions.submitEmail}
                onToggleEdit={accountData.actions.toggleEditEmail}
              />
              <ChangePassword
                confirmNewPassword={accountData.props.confirmNewPassword}
                currentPassword={accountData.props.currentPassword}
                editing={accountData.props.editingPassword}
                error={accountData.props.passwordError}
                newPassword={accountData.props.newPassword}
                onChangeConfirmNewPassword={
                  accountData.actions.changeConfirmNewPassword
                }
                onChangeCurrentPassword={
                  accountData.actions.changeCurrentPassword
                }
                onChangeNewPassword={accountData.actions.changeNewPassword}
                onSubmit={accountData.actions.submitPassword}
                onToggleEdit={accountData.actions.toggleEditPassword}
              />
              <div className="form-group">
                <H2>Delete Account</H2>
                <Spacer bottom="sm" />
                <P>Deleting your account cannot be undone.</P>
                {accountData.props.deleteError && (
                  <Spacer top="sm">
                    <Error>
                      <P>{accountData.props.deleteError}</P>
                    </Error>
                  </Spacer>
                )}
                <ModalPrompt
                  message="Please enter your password to confirm deletion of your account."
                  submitText="Delete Account"
                  cancelText="Cancel"
                  onSubmit={accountData.actions.deleteAccount}
                  inputType="password"
                >
                  {modal => (
                    <Button onClick={modal.actions.open}>Delete Account</Button>
                  )}
                </ModalPrompt>
              </div>
            </Tile>
          </Spacer>
        </React.Fragment>
      )}
    </AccountContainer>
  </StyleWrapper>
);

export default Account;
