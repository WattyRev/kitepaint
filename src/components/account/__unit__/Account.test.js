import React from "react";
import { shallow } from "enzyme";
import AccountContainer from "../../../containers/AccountContainer";
import Account from "../Account";

describe("Account", () => {
  it("renders", () => {
    const wrapper = shallow(<Account />);
    shallow(
      <div>
        {wrapper.find(AccountContainer).prop("children")({
          actions: {
            deleteAccount: jest.fn()
          },
          props: {
            user: {
              id: "abc-user"
            }
          }
        })}
      </div>
    );
  });
});
