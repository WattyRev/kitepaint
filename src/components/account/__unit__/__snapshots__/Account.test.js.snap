// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account renders 1`] = `
<styled.div>
  <Connect(AccountContainer)>
    <Component />
  </Connect(AccountContainer)>
</styled.div>
`;
